File: leetsync/2690-house-robber-iv/house-robber-iv.java
class Solution {
    public boolean isPossibleToRob(int[] nums, int k, int mid) {
        int n = nums.length;
        int house = 0;
        for (int i = 0; i < n; i++) {
            if (nums[i] <= mid) {
                house++;
                i++;// skipping adjacent house
            }
        }
        return house >= k;// we managed to rob at least k houses
    }

    public int minCapability(int[] nums, int k) {
        int n = nums.length;
        int left = Integer.MAX_VALUE, right = Integer.MIN_VALUE;
        for (int num : nums) {
            left = Math.min(left, num);
            right = Math.max(right, num);
        }
        int result = right;
        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (isPossibleToRob(nums, k, mid)) {
                result = mid;
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        return result;
    }
}

File: leetsync/2690-house-robber-iv/house-robber-iv.go
func minCapability(nums []int, k int) int {
	left := math.MaxInt32
	right := 0
	for _, num := range nums {
		left = min(left, num)
		right = max(right, num)
	}
	result := right
	for left <= right {
		mid := left + (right-left)/2
		if isPossibleToRob(nums, k, mid) {
			result = mid
			right = mid - 1
		} else {
			left = mid + 1
		}
	}
	return result
}
func isPossibleToRob(nums []int, k int, mid int) bool {
	n := len(nums)
	house := 0
	for i := 0; i < n; i++ {
		if nums[i] <= mid {
			house++
			i++
		}
	}
	return house >= k
}